pkg/kubelet/cm/cpumanager/cpu_manager.go

```diff
@@ -239,15 +239,61 @@ type reconciledContainer struct {
 	containerID   string
 }
 
-func (m *manager) reconcileState() (success []reconciledContainer, failure []reconciledContainer) {
+func (m *manager) removeStaleState() {
 	if !m.sourcesReady.AllReady() {
 		return
 	}
+
+	m.Lock()
+	defer m.Unlock()
+
+	activePods := m.activePods()
+	if len(activePods) == 0 {
+		// If there are no active pods, then we skip the removeal of stale
+		// state altogether. This ensures we don't remove something
+		// inadvertently and we will just try again next time this function is
+		// called.
+		return
+	}
+
+	activeContainers := make(map[string]struct{})
+	for _, pod := range activePods {
+		pstatus, ok := m.podStatusProvider.GetPodStatus(pod.UID)
+		if !ok {
+			// If even one pod does not have it's status set, we skip the
+			// removal of stale state altogether. This ensures we don't remove
+			// something inadvertently and we will just try again next time
+			// this function is called.
+			return
+		}
+		for _, container := range append(pod.Spec.InitContainers, pod.Spec.Containers...) {
+			containerID, err := findContainerIDByName(&pstatus, container.Name)
+			if err != nil {
+				// If even one container does not have it's containerID set, we
+				// skip the removal of stale state altogether. This ensures we
+				// don't remove something inadvertently and we will just try
+				// again next time this function is called.
+				return
+			}
+			activeContainers[containerID] = struct{}{}
+		}
+	}
+	for containerID := range m.state.GetCPUAssignments() {
+		if _, ok := activeContainers[containerID]; !ok {
```

Comment:
if activeContainers were mapped to a bool, I think you could just do "if activeContainers[containerID]"

Reply:
I had it that way to start with, but it doesn't actually work if there is an error (it panics).
It panics because when `containerID` is not present in the map, then `pod` is nil.

Reply:
Normally I would have fixed the bug in one commit, and then done the move to a new function in a second commit, but I forgot about this subtle change. Thanks for noting it.